### DISCUSSION TITLE
Applying ksp after evaluation

### DISCUSSION THREAD

**Discussion by jakoss** [slackhq/foundry#10] [Q&A] [answered]
I'm trying to replicate applying ksp conditionally, but i encountered a strange issue. I have something like that in my plugin (composite build):

```kotlin
target.afterEvaluate {
    if (myExtension.ksp) {
        target.pluginManager.apply("com.google.devtools.ksp")
        target.dependencies.add("ksp", project(":core:processor"))
    }
}
```
The issue is that the `:core:processor` does not run. There is completely nothing in logs, no errors or warnings. The ksp just does not generate code.

Doing the same code directly in target `build.gradle.kts` results with properly generated code. Is that something with `afterEvaluate`? Do i have to trigger processor compiler manually again? I cannot find anything like that in your plugin.

Anothing things, like dagger, implementation dependencies works fine, but the ksp is just not running for me.

I also tried to apply plugin in proper `build.gradle.kts` and then just add ksp dependencies in plugin. That does not work either. So it seems like doing `target.dependencies.add("ksp", project(":core:processor"))` in plugin does not trigger ksp.

UPDATE: Doing a SINGLE `target.dependencies.add("ksp", project(":core:processor"))` before the `afterEvaluate` triggers ksp even for ksp dependencies later on (so i can do single ksp before evaluation and then add other ksp depenendencies after that. But that still does not solve the issue for me, so i'm wondering how slack dealed with that

**Comment by ZacSweers** [comment, marked as answer]
We track whether or not the plugin is needed and then apply it IFF it hasn't been already applied.

The relevant functions are 
* "mark" functions like [this](https://github.com/slackhq/slack-gradle-plugin/blob/4849cf0f9f4d0d4b028946631ff0aa44a2cb3fc9/slack-plugin/src/main/kotlin/slack/gradle/SlackExtension.kt#L84). These track if we need to apply the plugin or no-op if already applied
* we look through our extension in afterEvaluate and configure KSP accordingly, like [this example with autoservice](https://github.com/slackhq/slack-gradle-plugin/blob/4849cf0f9f4d0d4b028946631ff0aa44a2cb3fc9/slack-plugin/src/main/kotlin/slack/gradle/SlackExtension.kt#L217-L222)


Anything beyond that though then I think you'll just need to make a repro project and file an issue on the KSP repo

**Reply by jakoss** [reply]
Ok thanks, I'll try to take this up with the ksp. Your approach is great and I learned a lot from the plugin
Thanks!
